assets/index.js

```diff
@@ -1,343 +1,255 @@
+	const context = canvas.getContext('2d')
+	context.drawImage(image, 0, 0)
+	const data = context.getImageData(0, 0, image.width, image.height).data
+
+	let brightnessSum = 0
+	for (let i = 0; i < data.length; i += 4) {
+		brightnessSum += (data[i] + data[i + 1] + data[i + 2]) / 3
+	}
+	return brightnessSum * 4 / data.length
 }
 
-function calculateBrightness(image, button) {
-    const canvas = document.createElement('canvas')
-    canvas.width = button.offsetWidth
-    canvas.height = button.offsetHeight
-
-    const context = canvas.getContext('2d')
-    const buttonRect = button.getBoundingClientRect()
-    context.drawImage(image, -buttonRect.left, -buttonRect.top)
-
-    const imageData = context.getImageData(0, 0, canvas.width, canvas.height)
-    const data = imageData.data
-
-    let brightnessSum = 0
-    let weightSum = 0
-
-    for (let i = 0; i < data.length; i += 4) {
-        const brightness = (data[i] + data[i + 1] + data[i + 2]) / 3
-        const weight = brightness / 255
-
-        brightnessSum += brightness * weight
-        weightSum += weight
-    }
-
-    return brightnessSum / weightSum
+function updateButtonColor(isBright) {
+	document.getElementById('random-button').classList.toggle('dark', isBright)
 }
 
-function isElementInViewport(element, threshold = 125) {
-    const rect = element.getBoundingClientRect()
-    const windowHeight = window.innerHeight || document.documentElement.clientHeight
-    const windowWidth = window.innerWidth || document.documentElement.clientWidth
-    const topVisible = rect.top + threshold < windowHeight
-    const bottomVisible = rect.bottom - threshold > 0
-    const leftVisible = rect.left + threshold < windowWidth
-    const rightVisible = rect.right - threshold > 0
-    return topVisible && bottomVisible && leftVisible && rightVisible
+(async function precalculateBrightness() {
+	await Promise.all(
+		screenshots.map(async screenshot => {
+			const image = new Image()
+			image.src = `./assets/img/screenshots/${screenshot}.webp`
+
+			await new Promise(resolve => {
+				image.onload = () => {
+					brightnessMap[screenshot] = calculateBrightness(image) >= 128
+					resolve()
+				}
+			})
+		})
+	)
+})()
+
+document.getElementById('random-button').addEventListener('click', getRandomScreenshot)
+getRandomScreenshot()
+
+function isInViewport(element, threshold = 125) {
+	const rect = element.getBoundingClientRect()
+	const windowHeight = window.innerHeight || document.documentElement.clientHeight
+	const windowWidth = window.innerWidth || document.documentElement.clientWidth
+	return (
+		rect.top + threshold < windowHeight &&
+		rect.bottom - threshold > 0 &&
+		rect.left + threshold < windowWidth &&
+		rect.right - threshold > 0
+	)
 }
 
 function handleScroll() {
-    articles.forEach((article) => {
-        if (isElementInViewport(article)) {
-            article.classList.add('animate')
-        }
-    })
+	for (const article of document.querySelectorAll('article')) {
```

Comment:
Ah, I see what you did there. I didn't thought about reducing DOM usage with each function call, especially considering how frequent `handleScroll` is called.

Reply:
It was a too obvious optimisation that I didn't thought about, because it was that simple, I just did remove immediate variables to reduce memory allocation. Sweet.